server/contrib/analyze.py

```diff
@@ -0,0 +1,345 @@
+    time2 = datetime.datetime.strptime(time2_string.split(".")[0], time_format)
+    time_difference = time2 - time1
+    return time_difference.total_seconds()
+
+def sort_by_client_time(backups):
+    """
+    Given a list of backups, sort by analytics time, mutating the original .
+    If a certains backup has no analytics; it is dropped.
+    """
+    def time_key(backup):
+        analytics = backup and backup.analytics()
+        if analytics:
+            time_string = analytics.get("time")
+            time_format = "%Y-%m-%d %X"
+            time = datetime.datetime.strptime(time_string.split(".")[0], time_format)
+            return time
+        return backup.created
+    backups.sort(key=time_key)
+
+### Diff Overview Generation ###
+def _get_backup_range(backups, backup_id, bound):
+    """
+    Naive Implementation
+    Returns a dictionary: {
+        "backups": backups,
+        "prev_backup_id": prev_backup_id,
+        "backup_id": backup_id
+        "next_backup_id": next_backup_id
+    }
+    backups = unique backups `bound` away from backup with `backup_id`
+    prev/next_backup_id = prev/next backup_id not part of backups
+    """
+    unique_backups_tuples, id_map, index_map = _get_unique_backups(backups)
+    unique_backups = [backup_tuple[0] for backup_tuple in unique_backups_tuples]
+    # invalid backup_id
+    if backup_id not in id_map:
+        return {}
+
+    # find included backup_id
+    backup_id = id_map[backup_id]
+
+    backup_index = index_map[backup_id]
+
+    # get prev and curr ids
+    prev_backup_id = unique_backups[max(0, backup_index-bound-1)].id
+    next_backup_id = unique_backups[min(len(unique_backups)-1, backup_index+bound+1)].id
+
+    unique_backups = unique_backups[max(0, backup_index-bound-1)\
+                                    :min(len(unique_backups), backup_index+bound+1)]
+    return {
+        "backups": unique_backups,
+        "prev_backup_id": prev_backup_id,
+        "backup_id": backup_id,
+        "next_backup_id": next_backup_id
+    }
+
+def _recent_backup_finder(backups):
+    """
+    Given list of backups, creates a HOF used for finding the most recent backup.
+    """
+    partner_index = 0
+    backup_count = len(backups)
+    def _get_recent_backup(current_analytics):
+        """
+        `current_analytics`: contains the time to find the most recent backup
+        attempts to find a backup b_p such that
+            "b_p.time" is the largest possible AND
+            "b_p.time" < `current_time`
+        Returns backup b_p or None if nothing is found
+        """
+        nonlocal partner_index
+        while partner_index < backup_count:
+            backup = backups[partner_index]
+
+            analytics = backup and backup.analytics()
+            if analytics:
+                time_diff = _get_time_diff_seconds(analytics, current_analytics)
+                if time_diff < 0: # curr_analytics - analyics < 0 => too far ahead
```

Comment:
I dont think this consecutive `if` can be joined without making the code more complicated.
```
if time_diff < 0 and partner_index > 0: # curr_analytics - analyics < 0 => too far ahead
    prev_backup = backups[partner_index-1]
    return prev_backup
elif time_diff < 0:
    return None
```